### ISSUE TITLE
x-height calculation incorrect in calibre viewer on windows with 2.6.0

### ISSUE THREAD

**Issue by kovidgoyal** [mathjax/MathJax#1422]
I recently updated the version of mathjax included with calibre to 2.6.0. However, on windows it calculates the font x-height incorrectly. linux and OS X are fine. 

I tried updating to 2.6.1 -- no change.

Note that calibre uses Qt WebKit to render HTML and if you want to play with the mathjax bundled with calibre you can find it in resources/viewer/mathjax in the calibre installation directory. It is unbundled, so it should be easy to modify and edit. A sample epub file which shows this problem is https://manual.calibre-ebook.com/_static/mathjax.epub

Simply open it in the calibre viewer and you will see that math is getting typeset at approx 50% of correct height. You can right click in the viewer and choose Inspect to get access to a javascript console. 

I did a bit of debugging and I one difference I found is that the height of the ExSpan ins ~550px on linux and 300px on windows. However, changing .MathJax_SVG_ExBox to have height 120ex instead of 60ex does not affect rendering, so I'm not sure how to proceed. If you can tell me what I need to change to affect the x height calculation, I can probably special case it for calibre on windows.

**Comment by kovidgoyal** [comment]
Oh and I forgot to mention, calibre uses SVG output, so the ExSpan I am talking about is in SVG/jax.js

**Comment by pkra** [comment]
What Calibre version would this be happening on? Also, what QTWebKit version are you using?

FWIW, QTWebKit has been problematic in the past since it's a rather stripped down version of WebKit so it might be hard to fix (unless you're already on the Chromium-based version, I suppose; that might have a new set of problems we haven't seen yet).

**Comment by pkra** [comment]
One thing worth checking might be http://docs.mathjax.org/en/latest/output.html#viewport-meta-tag, depending on whether Calibre does that or not (the epub file does not have any meta tags for viewport size).

**Comment by kovidgoyal** [comment]
Qt 5.5.1 and Qt WebKit (the old chromium engine), not Qt WebEngine. (Although that will likely change in the future, when I migrate the viewer to Qt WebEngine). 

I dont really need you to fix anything, this is almost certainly a bug in Qt WebKit (the ex unit is basically much smaller in windows than on other platforms. If you can tell me where MathJax does its x height calculations for SVG output, I will simply special case them in the version bundled with calibre to workaround this bug in Qt WebKit.

**Comment by kovidgoyal** [comment]
calibre version 2.53 (the latest)

**Comment by dpvc** [comment]
The ex-height is computed on [line 251](https://github.com/mathjax/MathJax/blob/master/unpacked/jax/output/SVG/jax.js#L251) of the unpacked/jax/output/SVG/jax.js file.

**Comment by kovidgoyal** [comment]
Yeah, I tried changing it there already, it did not work. The rendered height remains incorrect. I changed 60 to 30 which changed the value of ex from 5 to 10 but it does not seem to affect actual rendering. You can try it for yourself, as I outlined in my first post.

**Comment by dpvc** [comment]
OK, it turns out that the value determined in line 251 is used internally to handle units in terms of ex (e.g., \hskip 3ex), but the size of the actual SVG element is not based on that.  It is given explicitly in ex, so if the ex-height of the font is not actually correct, it won't match well.

Fortunately, the measurements all go through the `Ex()` method (and nothing else currently uses that), so you can modify [line 489](https://github.com/mathjax/MathJax/blob/master/unpacked/jax/output/SVG/jax.js#L489) and that should allow you to adjust the size of the final SVG result.  See if that helps.

**Comment by kovidgoyal** [comment]
That does not quite work either, it seems to have the same effect as setting scale in the SVG output plugin options. While this works great for changing the x-height, it causes the alignment of equations to be incorrect. The width of the equation box has changed but the margin calculation is not changed, so the equation is no longer centered as it should be. 

This is looks like a bug in how scale is implemented -- so you should probably investigate fixing that. 
If you can tell me where margins are calculated for equation boxes, I can probably look into creating a patch that should allow me to just use scale n windows to workaround the problem.

**Comment by dpvc** [comment]
> That does not quite work either, it seems to have the same effect as setting scale in the SVG output plugin options. While this works great for changing the x-height, it causes the alignment of equations to be incorrect. The width of the equation box has changed but the margin calculation is not changed, so the equation is no longer centered as it should be.

I am not able to reproduce that, and I'm not sure I understand.  The margins are set using the same `Ex()` routine, so any change you make there should affect the margins and the vertical-align, as well as the width and height.  They should all scale at the same rate, so the alignment should be maintained.  That is certainly how it works for me (but I'm not using your Windows environment where the ex-height is wrong to start with).  But I don't really understand how the margins would not be changed.

> This is looks like a bug in how scale is implemented -- so you should probably investigate fixing that. 

We have no other reports of this behavior, and I can't reproduce it, so I'm wondering if it not another Qt WebKit issue.  HOw far off is the alignment?  Is it more than one pixel?  Can you provide some screen shots?

> If you can tell me where margins are calculated for equation boxes, I can probably look into creating a patch that should allow me to just use scale n windows to workaround the problem.

The styles are set at [lines 2081 through 2091](https://github.com/mathjax/MathJax/blob/master/unpacked/jax/output/SVG/jax.js#L2081-L2091), with some possible adjustments following that.  It may be that the rounding to pixels part is the issue for you, as it is based on the measured ex-height.  Have you retained your changes to that as well (the ones at line 251)?

**Comment by kovidgoyal** [comment]
![100](https://cloud.githubusercontent.com/assets/1308621/14060077/cb8aee62-f37d-11e5-8ea7-298d72d3b077.png)
![200](https://cloud.githubusercontent.com/assets/1308621/14060078/cb8b3390-f37d-11e5-894d-5edfa6e0f57c.png)

Screenshots -- the only difference between the two is one is with scale 100 aand the other is with scale 200. The screen shots are from windows, but the same thing happens on linux, namely the equations are no longer centered properly.

**Comment by kovidgoyal** [comment]
And note that the above screenshot are with completely vanilla MathJax 2.6.1 with no changes from me.

**Comment by dpvc** [comment]
Thanks for the screen shots.  When an expression has a tag (e.g., an equation number), then the SVG output must be made the full width of the container.  It turns out that when scaling is in effect, the container is also scaled, and so the SVG is being made too wide (and so extends off the right-hand side of the screen).

This can be fixed by changing [line 252](https://github.com/mathjax/MathJax/blob/master/unpacked/jax/output/SVG/jax.js#L252) to be 

```
        cwidth = div.previousSibling.firstChild.offsetWidth / this.config.scale * 100;
```

See if that resolves the issue for you.

**Comment by kovidgoyal** [comment]
That does indeed fix it, thanks.

**Comment by kovidgoyal** [comment]
The only remaining problem I can see is that on linux and os x the equation numbers are not visible they are off beyond the right side of the page. Any tips on where to look to adjust their positioning?

This si not a big deal, at least, I have had no users complaining about it, but since I am here, it would be nice to fix.

**Comment by dpvc** [comment]
> on linux and os x the equation numbers are not visible they are off beyond the right side of the page.

Is that with an unmodified MathJax?  Is it with scale = 100% , or with a larger scale?

The value of `cwidth` in my message above is what determines the width of the SVG element when equation numbers or tags are used (but it is converted to em's on [line 264](https://github.com/mathjax/MathJax/blob/master/unpacked/jax/output/SVG/jax.js#L264), and the em-size is determined from the ex-size in the previous line, so if that isn't being computed correctly, it will affect the final value of `cwidth`).  So if the labels are appearing off screen, then the value of `cwidth` my be too big.  We do not have any report of this occurring, and I have not been able to reproduce it in browsers on OS X.

**Comment by kovidgoyal** [comment]
Unmodified mathjax with scale 100

**Comment by dpvc** [comment]
> on linux and os x the equation numbers are not visible they are off beyond the right side of the page.

Is this for equations that are wrapped, or all equations (even short ones)?

**Comment by kovidgoyal** [comment]
All equations, as far as I can tell, it is a constant offset, independent of equation size.

**Comment by dpvc** [comment]
==> Merged (fix for SVG width issue).

**Comment by dpvc** [comment]
==> In testsuite

MathMLToDisplay/issue1422.html
